### PR TITLE
board_specific_demo/ameba_secure_world_demo: Add secure_world example

### PR DESCRIPTION
**Pull request by Tirtha-Samsung**
Signed-off-by: Tirtha Biswas <biswas.t@samsung.com>

### CODE REVIEW

**Comment by Tirtha-Samsung** [comment]
```
TASH>>secure_world -r

TASH>>

Security Fault: 

Secure State: 1



Stacke 

R0 = 0x00000072

R1 = 0x00000072

R2 = 0x0e099583

 = 0x20095dd7

R12 = 0x00000000

LR = 0x0e08cb7d

PC = 0x0e00b93a

PSR = 0x61000000



Current: 

EXC_RETU = 0xffffffbd

MSP = 0x20095d70

PSP = 0x00000000

xPSR 0xa0000007

CFSR  = 0x00000000

HFSR  = 0x00000000

DFSR= 0x00000000

MMFAR = 0x00000000

BFAR  = 0x00000000

AF  = 0x00000000

PriMask = 0x00000000

SVC priority: 0x00

PendSVC priority: 0x00

Systick priority: 0x00

MSP_NS = 00095d70

PSP_NS = 0x00000000

CFSR_NS  = 0x00000000

HFSNS  = 0x00000000

DFSR_NS  = 0x00000000

MMFAR_NS = 0x00000000

BFAR_NS  = 0x00000000

AFSR_NS  = 0x00000000

SVC iority NS: 0x00

PendSVC priority NS: 0xe0

Systick priori NS: 0x80


```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/Make.defs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+############################################################################
+# apps/examples/board_specific_demo/ameba_secure_storage_demo/Make.defs
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/Make.defs:19]: This needs to be removed please

File: apps/examples/board_specific_demo/ameba_secure_world_demo/Make.defs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# Adds selected applications to apps/ build
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/Make.defs:20]: We don't need the Nuttx header at all since it's a new file.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/Makefile [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+############################################################################
+# apps/examples/board_specific_demo/ameba_secure_world_demo/Makefile
+#
+#   Copyright (C) 2008, 2010-2013 Gregory Nutt. All rights reserved.
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/Makefile:21]: Same here, we haven't ported this from Nuttx so we don't need the header I think.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+examples/secure_world
+^^^^^^^^^^^^^^
+
+  Secure World APIs is used to check when we access to read or to write in non-secure world, 
+  it should be crashed.
+I
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt:6]: Could you remove this stray line please and add the usage description here?

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+{
+	int option;
+	volatile int address;
+	while (option = getopt(argc, argv, "rw")) {
+		switch (option) {
+			case 'r':
+				address = *ADDR;
+				break;
+			case 'w':
+
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:54]: Could you remove this extra line please?

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <security/security_common.h>
+#include <security/security_api.h>
+#include <tinyara/seclink.h>
+#include <tinyara/security_hal.h>
+#include <security/security_ss.h>
+#define SHOW_USAGE printf("Usage: secure_world -r -w\n");
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:30]: Could you please modify SHOW_USAGE as
Usage: secure_world -r -w
r : run the read test
w : run the write test

And add this usage description to the README doc too

**Comment by amogh-samsung** [comment]
@sunghan-chang we have confirmed that raw access to the secure world for reading or writing causes crash
The app reproduces the same

Please let us know if the functionality is as per requirements
Cosmetic changes will be taken care of

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+			case 'r':
+				address = *ADDR;
+				break;
+			case 'w':
+
+				*ADDR = 1234; 
+				printf("%d",*ADDR);
+				break;
+			default:
+				SHOW_USAGE
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:59]: Could you please add some error message like "invalid input" before redirecting the user to the usage.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+#else
+int secure_world_main(int argc, char *argv[])
+#endif
+{
+	int option;
+	volatile int address;
+	while (option = getopt(argc, argv, "rw")) {
+		switch (option) {
+			case 'r':
+				address = *ADDR;
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:51]: Could you rename address to value? I feel like the name address is not appropriate here

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <security/security_common.h>
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:25]: The includes from here on are not related to this. Could you please check and remove them.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+  This shows how to call write and read from Secure Storage with SE APIs,
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt:8]: This line is not related to the example added. Please remove it.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt [hunk 11]
```diff
@@ -0,0 +1,13 @@
+examples/secure_world
+^^^^^^^^^^^^^^
+
+  Secure World APIs is used to check when we access to read or to write in non-secure world,
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/README.txt:4]: We are currently not using any APIs, so please change the description. I think something like "This example depicts how the access from non secure to secure world is restricted. When non secure world tried to access(write/read) from the secure world region, it causes a secure fault"

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+#include <security/security_api.h>
+#include <tinyara/seclink.h>
+#include <tinyara/security_hal.h>
+#include <security/security_ss.h>
+#define SHOW_USAGE printf("Usage: secure_world -r -w\n");
+
+#ifdef  CONFIG_AMEBALITE_TRUSTZONE
+#define ADDR ((volatile uint32_t*)0x735C021)
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:33]: How about getting the addr with -a? If no input for addr, use this value as default. And help need to show valid range.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	volatile int address;
+	while (option = getopt(argc, argv, "rw")) {
+		switch (option) {
+			case 'r':
+				address = *ADDR;
+				break;
+			case 'w':
+
+				*ADDR = 1234; 
+				printf("%d",*ADDR);
```
Comment (jeongarmy) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:56]: Could you add logs about test like `Write %d to XXXX`?

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c (hunk 8 above)
Comment (jeongarmy) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:51]: Could you add logs about test like `Try to read addr 0xXXXX`?

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+int main(int argc, FAR char *argv[])
+#else
+int secure_world_main(int argc, char *argv[])
+#endif
+{
+	int option;
+	volatile int value;
+	while (option = getopt(argc, argv, "rw")) {
+		switch (option) {
+			case 'r':
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:47]: By TizenRT coding rules, `switch` and `case` should be the same indentation.
```
        switch (option) {
        case 'r':
```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+				value = *ADDR;
+				break;
+			case 'w':
+				*ADDR = 1234; 
+				printf("%d",*ADDR);
+				break;
+			default:
+				printf("INVALID INPUT");
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:55]: Need LF `\n` at end of the string.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c (hunk 13 above)
Comment (sunghan-chang) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:52]: Is this a meaningful code? Because it will be in assert, I think printing after the code will not be executed.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int secure_world_main(int argc, char *argv[])
+#endif
+{
+	int option;
+	volatile int value;
+	while ((option = getopt(argc, argv, "a:rw"))!=-1) {
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:47]: `) != -1`
need a space around `!=`

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+		switch (option) {
+		
+		case 'a':
+			sscanf(optarg,"%x",&address);
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:51]: need a space after `,`
```
", &add
```
Please check all of the same violation.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+#include <stdio.h>
+
+#define SHOW_USAGE printf("Usage: secure_world -r -w -a\n \
+		r: run the read test\n \
+		w: run the write test\n \
+		a: take the address as input\n \
+		Valid Range: 0x7034C020 - 0x7039C000\n");
+
+#ifdef  CONFIG_AMEBALITE_TRUSTZONE
+volatile uint32_t* address = 0x735C021;
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:32]: `uint32_t *address`

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+
+#define SHOW_USAGE printf("Usage: secure_world -r -w -a\n \
+		r: run the read test\n \
+		w: run the write test\n \
+		a: take the address as input\n \
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:28]: This is optional. Let's explain it.
And leave an option among r or w  is mandatory.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+			sscanf(optarg,"%x",&address);
+			break;
+		case 'r':
+			printf("Trying to read data from %x\n",address);
+			value = *address;
+			break;
+		case 'w':
+			printf("Trying to write data at %x\n",address);
+			*address = 1234; 
+			break;
```
Comment (abhishek-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:60]: this will not work if we give ```secure_world -r -a 0x7....``` as we process -r flag first and it crashes with the default address.  Also, it doesn't check if one of -r or -w flags are given as they are mandatory. So we better set some variables to process the flags and then access the memory. 

Example : 

```
bool r_flag = false;
bool w_flag = false;
.
.
if (!r_flag && !w_flag) SHOW_USAGE
```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c (hunk 19 above)
Comment (amogh-samsung) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:28]: Since -a should be followed by an input
Could you depict that in the show_usage function too?
"... -a Secure_World_Address" or something like that

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+			sscanf(optarg, "%x", &address);
+			break;
+		case 'r':
+			printf("Trying to read data from %x\n", address);
+			value = *address;
+			break;
+		case 'w':
+			printf("Trying to write data at %x\n", address);
+			*address = 1234; 
+			break;
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:59]: To make sure that either -r or -w is selected and the operation is done on the address specified with -a by the user,
You could set a boolean flag as Abhishek suggested to mark if Read/Write option has been chosen.
Then perform the operation later on the input address as per the selected option.

Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:59]: To make sure that either -r or -w is selected and the operation is done on the address specified with -a by the user,
You could set a boolean flag as Abhishek suggested to mark if Read/Write option has been chosen.
Then perform the operation later on the input address as per the selected option.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+
+#define SHOW_USAGE printf("Usage: secure_world -r -w -a\n \
```
Comment (amogh-samsung) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:25]: The -a option allows us to generalize the example to work on different boards for their respective secure world regions.
So what is no address input is given, is there a default address that the test works for?

Also, can we add the Secure World regions of tp1e, etc to this usage help output so that the user can test conveniently?
Later, we can add the addresses of other boards too as and when needed so the user can print the help menu and then give the right input for testing.

Comment (amogh-samsung) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c]: Take this into consideration please, it helps future proof the application.

We can make -a mandatory if we provide info to the user in the help menu about Secure world address ranges for testing.
The default values used currently would only work for Tp1e and perhaps not if the test was run on another board.

@sunghan-chang @abhishek-samsung  your opinion please?

Comment (abhishek-samsung) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c]: I don't think we need to support other boards as this is a board-specific example. We can add the appropriate default addresses based on configs like the secure storage example.

https://github.com/Samsung/TizenRT/blob/eca807f0c25b342b12b233f6fd6e3e63c0e7c4a9/apps/examples/board_specific_demo/ameba_secure_storage_demo/secure_storage.c#L46-L53

It will be easier for the user to execute the command without providing the address, so I think it's better to have a default address based on the board tp1x/tp1e.

> Also, can we add the Secure World regions of tp1e, etc to this usage help output so that the user can test conveniently?

I agree with this, we can add the region addresses in help according to the board tp1x/tp1e

**Comment by Tirtha-Samsung** [comment]
"""
TASH>>secure_world -a 0x7035C025

TASH>>Usage: secure_world -a secure_world_address -r/-w

        -r: run the read test

                -w: run the write test

                -a sece_world_address: take the address as input

                Valid Range: 0034C020 - 0x7039C000

                The default address when no input adess is given is 0x735C021

TASH>>

TASH>>

TASH>>

TASH>>secure_world -r           

TASH>>Trying to read data from 7035c025



Security Fault: 

Sere State: 1



Stacked: 

R0 = 0x00000022

R1 = 0x2008f0

R2 = 0x00000000

R3 = 0x7035c025

R12 = 0x00041430

R = 0x0e00b987

PC = 0x0e00b988

PSR = 0x61000000



Cuent: 

EXC_RETURN = 0xffffffbd

MSP = 0x20095d68

PSP = 00000000

xPSR = 0xa0000007

CFSR  = 0x00000000

HFSR  =x00000000

DFSR  = 0x00000000

MMFAR = 0x00000000

BFAR  0x00000000

AFSR  = 0x00000000

PriMask = 0x00000000

SVC priority: 0x00

PendSVC priority: 0x00

Systick priorit 0x00

MSP_NS = 0x20095d68

PSP_NS = 0x00000000

CFSR_NS= 0x00000000

HFSR_NS  = 0x00000000

MMFAR_NS = 0x00000000

BFAR_NS  = 0x00000000

AFSR_NS  =x00000000

SVC priority NS: 0x00

PendSVC priority NS: 0xe

Systick priority NS: 0x80


"""

**Comment by Tirtha-Samsung** [comment]
```
TASH>>secure_world -r -a 0x7035C025

TASH>>Trying to read data from 7035c025



Security Fau: 

Secure State: 1



Stacked: 

R0 = 0x00000022

R1 0x2008fbf0

R2 = 0x00000000

R3 = 0x7035c025

R12 = 0x041430

LR = 0x0e00b987

PC = 0x0e00b988

PSR = 0x6100000



Current: 

EXC_RETURN = 0xffffffbd

PSP = 0x00000000

xPSR = 0xa0000007

HFSR  = 0x00000000

DFSR  = 0x00000000

MMFAR = 0x0000000

BFAR  = 0x00000000

AFSR  = 0x00000000

PriMask = 0x00000000

SVC priority: 0x00

PendSVC priority: 0x00

Systi priority: 0x00

MSP_NS = 0x20095d50

CFSR_NS  = 0x00000000

HFSR_NS  = 0x00000000

DFSR_NS  = 00000000

MMFAR_NS = 0x00000000

BFAR_NS  = 0x00000000

SR_NS  = 0x00000000

SVC priority NS: 0x00

PendSVC priori NS: 0xe0

Systick priority NS: 0x80

```

**Comment by Tirtha-Samsung** [comment]
```
TASH>>secure_world -w

TASH>>Trying to write data at 735c021



Security Fault

Secure State: 1



Stacked: 

R0 = 0x00000020

R1 =x2008fbf0

R2 = 0x000004d2

R3 = 0x0735c021

R12 = 0x000430

LR = 0x0e00b93d

PC = 0x0e00b942

PSR = 0x61000000

Current: 

EXC_RETURN = 0xffffffbd

MSP = 0x20095d68

PSP = 0x00000000

xPSR = 0xa0000007

CFSR  = 0x00000000

FSR  = 0x00000000

DFSR  = 0x00000000

BFAR  = 0x00000000

AFSR  = 0x00000000

PriMask = 0x00000000

SVC priority: 0x00

PendSVC priority: 0x00

Systickriority: 0x00

MSP_NS = 0x20095d68

PSP_NS = 0x00000000

FSR_NS  = 0x00000000

HFSR_NS  = 0x00000000

DFSR_NS  = 0x000000

MMFAR_NS = 0x00000000

BFAR_NS  = 0x00000000

AF_NS  = 0x00000000

SVC priority NS: 0x00

PendSVC priorityS: 0xe0

Systick priority NS: 0x80


```

**Comment by Tirtha-Samsung** [comment]
```
ASH>>secure_world -a 0x7035C025 -w

TASH>>Trying to write data at 7035c025



Security Faul 

Secure State: 1



Stacked: 

R0 = 0x00000021

R1 0x2008fbf0

R2 = 0x000004d2

R3 = 0x7035c025

R12 = 0x001430

LR = 0x0e00b93d

PC = 0x0e00b942

PSR = 0x61000000

Current: 

EXC_RETURN = 0xffffffbd

PSP = 0x00000000

xPSR = 0xa0000007

CFSR  = 0x00000000

HFSR  = 0x00000000

DFSR  = 0x00000000

BFAR  = 0x00000000

AFSR  = 0x00000000

PriMask = 0x00000000

SVC priority: 0x00

PendSVC priority: 0x00

Systicpriority: 0x00

MSP_NS = 0x20095d50

PSP_NS = 0x00000000

CFSR_NS  = 0x00000000

HFSR_NS  = 0x00000000

DFSR_NS  = 00000000

MMFAR_NS = 0x00000000

BFAR_NS  = 0x00000000

AR_NS  = 0x00000000

SVC priority NS: 0x00

PendSVC prioritNS: 0xe0

Systick priority NS: 0x80

'''

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+int secure_world_main(int argc, char *argv[])
+#endif
+{
+	int option;
+	volatile int value;
+	bool r_flag = false, w_flag = false;
+	while ((option = getopt(argc, argv, "a:rw")) != -1) {
+		switch (option) {
+		case 'a':
+			sscanf(optarg, "%x", &address);
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:53]: Do we need to use `sscanf` instead of value assign?
```
&address = (uint32_t *)optarg;
or
address = (uint32_t *)optarg;
```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+		case 'w':
+			w_flag = true;
+			break;
+		default:
+			printf("INVALID INPUT\n");
+			SHOW_USAGE
+			return 0;
+		}
+	}
+	if (!r_flag && !w_flag)
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:67]: 1. Need brackets even it's a single statement
2. closing and else should be at the same line
```
if (...) {
   do
} else if {
} else {
}
```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int secure_world_main(int argc, char *argv[])
+#endif
+{
+	int option;
+	volatile int value;
+	bool r_flag = false, w_flag = false;
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:49]: bool r_flag = false;
bool w_flag = false;

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:23]: Need to include several headers
```
#include <tinyara/config.h>   for CONFIG_XX
#include <stdio.h>    for printf
#include <stdint.h>   for uint32_t
#include <unistd.h>   for getopt
#include <stdbool.h>  for bool and false, true
```

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+
+#define SHOW_USAGE printf("Usage: secure_world [opt2] secure_world_address [opt1]\n \
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:25]: In Ubuntu, mandatory option is not in square brackets.
```
Usage: secure_world OPT1 [OPT2]
    OPT1
          -r:
          -w: .....
    OPT2
          -a ADDRESS: xxxxxx

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+		opt1(-r/-w): One of the option is mandatory\n \
+		-r: run the read test\n \
+		-w: run the write test\n \
+		opt2(-a):\n \
+	        -a: secure_world_address: take the address as input, default: 0x735C021\n \
+		Valid Range: 0x7034C020 - 0x7039C000\n \");
+
+#ifdef  CONFIG_AMEBALITE_TRUSTZONE
+volatile uint32_t  *address = 0x735C021;
+#endif
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:35]: How about adding it in Kconfig because it's a different value for every board?
If we add it as Kconfig, we can keep the same code for every board. (No need to modify the code itself)

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c (hunk 22 above)
Comment (sunghan-chang) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c]: How about adding it in Kconfig because it's a different value for every board?
If we add it as Kconfig, we can keep the same code for every board. (No need to modify the code itself)

File: apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c (hunk 23 above)
Comment (abhishek-samsung) [reply apps/examples/board_specific_demo/ameba_secure_world_demo/secure_world.c:54]:
```
#include <[unistd.h](https://linux.die.net/include/unistd.h)>

int getopt(int argc, char * const argv[],
           const char *optstring);

extern char *optarg;
extern int optind, opterr, optopt;
```
Since optarg is of type char, I think we cant assign it.

File: apps/examples/board_specific_demo/ameba_secure_world_demo/Kconfig [hunk 29]
```diff
@@ -0,0 +1,15 @@
+#
```
Comment (sangam-swami) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/Kconfig:1]: you can add test results to commit message instead of conversation

File: os/board/rtl8720e/Kconfig [hunk 30]
```diff
@@ -47,6 +47,12 @@ config AMEBALITE_FLASH_BLOCK_SIZE
     help
         Setting this vaule to configure erase unit size of esp32 SPI flash.
 	
+config AMEBALITE_SECURE_WORLD_EXAMPLE
```
Comment (sangam-swami) [inline os/board/rtl8720e/Kconfig:50]: you can use AMEBALITE_SECURE_WORLD_ADDRESS?

File: apps/examples/board_specific_demo/ameba_secure_world_demo/Kconfig [hunk 31]
```diff
@@ -0,0 +1,15 @@
...
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_SECURE_WORLD
+	bool "Secure World example"
+	default n
+	depends on ARMV8M_TRUSTZONE && SE_AMEBA
```
Comment (sangam-swami) [inline apps/examples/board_specific_demo/ameba_secure_world_demo/Kconfig:9]: I think the dependency should be on AMEBAD_TRUSTZONE || AMEBALITE_TRUSTZONE ?

**Review by sunghan-chang — APPROVED** [review]
